packages/core/src/electron-main/electron-application.ts

```diff
@@ -0,0 +1,440 @@
+                    width: bounds.width,
+                    height: bounds.height,
+                    x: bounds.x,
+                    y: bounds.y
+                });
+            } catch (e) {
+                console.error('Error while saving window state:', e);
+            }
+        };
+        let delayedSaveTimeout: NodeJS.Timer | undefined;
+        const saveWindowStateDelayed = () => {
+            if (delayedSaveTimeout) {
+                clearTimeout(delayedSaveTimeout);
+            }
+            delayedSaveTimeout = setTimeout(saveWindowState, 1000);
+        };
+        electronWindow.on('close', saveWindowState);
+        electronWindow.on('resize', saveWindowStateDelayed);
+        electronWindow.on('move', saveWindowStateDelayed);
+    }
+
+    /**
+     * Catch window closing event and display a confirmation window.
+     */
+    protected attachWillPreventUnload(electronWindow: BrowserWindow): void {
+        // Fired when a beforeunload handler tries to prevent the page unloading
+        electronWindow.webContents.on('will-prevent-unload', event => {
+            const preventStop = 0 !== dialog.showMessageBox(electronWindow, {
+                type: 'question',
+                buttons: ['Yes', 'No'],
+                title: 'Confirm',
+                message: 'Are you sure you want to quit?',
+                detail: 'Any unsaved changes will not be saved.'
+            });
+
+            if (!preventStop) {
+                // This ignores the beforeunload callback, allowing the page to unload
+                event.preventDefault();
+            }
+        });
+    }
+
+    /**
+     * Catch certain keybindings to prevent reloading the window using keyboard shortcuts.
+     */
+    protected attachGlobalShortcuts(electronWindow: BrowserWindow): void {
+        if (this.config.electron?.disallowReloadKeybinding) {
+            const accelerators = ['CmdOrCtrl+R', 'F5'];
+            electronWindow.on('focus', () => {
+                for (const accelerator of accelerators) {
+                    globalShortcut.register(accelerator, () => { });
+                }
+            });
+            electronWindow.on('blur', () => {
+                for (const accelerator of accelerators) {
+                    globalShortcut.unregister(accelerator);
+                }
+            });
+        }
+    }
+
+    /**
+     * Start the NodeJS backend server.
+     *
+     * @return Running server's port promise.
+     */
+    protected async startBackend(): Promise<number> {
+        // Check if we should run everything as one process.
+        const noBackendFork = process.argv.indexOf('--no-cluster') !== -1;
+        // Any flag/argument passed after `--` will be forwarded to the backend process.
+        const backendArgvMarkerIndex = process.argv.indexOf('--');
+        const backendArgv = backendArgvMarkerIndex === -1 ? [] : process.argv.slice(backendArgvMarkerIndex + 1);
```

Comment:
This is new. It took some time to figure out why the VS Code extensions do no start when I launch the app from the sources. Is there a chance to have it as before? This is a breaking change for all our clients for sure.

Reply:
The reasoning here is that app makers will most likely create their own cli, which ought to be more high level than what the backend cli expects. I think a breakage would be justified here, else we would have to defer the responsibility of re-routing command line arguments onto app makers.

Reply:
>  I think a breakage would be justified here,

Yes, you're right but can we include it in a separate PR when it works? I do not understand why you want to put this CLI change to this PR. It was about DI and not changing the way we handle the CLI arguments.

> that app makers will most likely create their own cli

Do you have a real use-case where you want to use these new features? If yes, would you mind sharing them? If no, let's wait till we have a real need and do not make premature architectural decisions that require a breaking change.

Reply:
> Yes, you're right but can we include it in a separate PR when it works?

I'm ok with this, let's do that.

> I do not understand why you want to put this CLI change to this PR. It was about DI and not changing the way we handle the CLI arguments.

While doing so I ended up breaking apart the electron-main thing we had before which was really hard to look at. Might have been carried away.

> Do you have a real use-case where you want to use these new features?

Yes, we are making an electron app at Ericsson and already worked around having DI to handle the app's CLI.

Am ok to do that later.

Reply:
Reverted this change.